### PR TITLE
perf: router workers and optimizations

### PR DESCRIPTION
**Pull request by DanielSchiavini**
- Fixes https://github.com/curvefi/curve-js/issues/312
- Runs the router graph creation and route finding algorithms in **web workers, for UI responsivity**
- This should be backwards compatible for any clients. It **runs synchronously in nodejs**.
- **Route graph** algorithm is not changed, just moved (a few changes were done to avoid list spread syntax, tsc uses helpers for that that do not work in a web worker)
- Route finding algorithm is optimized to **preemptively stop routes** when they are not possibly the best choices based on minTvl and route length.
  - Route finding is roughly 3x faster (worst case, no match) **up to 300x faster** (best case, direct match) - check yourself if you wish with commit https://github.com/curvefi/curve-js/commit/892ec4e1c8c5fdfe84ba4dce325233e880ceb08a

### CODE REVIEW

**Comment by DanielSchiavini** [comment]
Note in the following profile how the arced part of the code is now being ran in a worker.
image

The route graph construction still happens in the UI thread for now, as we need the global curvejs state.

**Comment by DanielSchiavini** [comment]
The new algorithm was first isolated in https://github.com/curvefi/curve-js/pull/386 so it might be easier to review there. In that PR the code was 3x faster, but commit https://github.com/curvefi/curve-js/pull/385/commits/bad8491cb8a713080b31d488b8bcc65df84685c3 made the best case dramatically more efficient

File: tsconfig.json [hunk 1]
```diff
@@ -4,9 +4,11 @@
 
     /* Basic Options */
     // "incremental": true,                         /* Enable incremental compilation */
-    "target": "es5",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (DanielSchiavini) [inline tsconfig.json:7]: This makes the worker code output a bit better - I think es6 is already well supported anyway

**Comment by DanielSchiavini** [comment]
[Screencast from 26-08-24 09:33:15.webm](https://github.com/user-attachments/assets/8a358830-0636-44bc-869a-976f71dc95da)

File: src/router.ts [hunk 2]
```diff
@@ -1,459 +1,85 @@
...
-                    if (!poolData?.is_lending && _isVisitedPool(step.poolId, route)) continue;
-
-                    // 4 --> 6, 5 --> 7 not allowed
-                    // 4 --> 7, 5 --> 6 allowed
-                    const routePoolIdsPlusSwapType = route.route.map((s) => s.poolId + "+" + _handleSwapType(s.swapParams[2]));
-                    if (routePoolIdsPlusSwapType.includes(step.poolId + "+" + _handleSwapType(step.swapParams[2]))) continue;
-
-                    const poolCoins = poolData ? poolData.wrapped_coin_addresses.concat(poolData.underlying_coin_addresses) : [];
-                    // Exclude such cases as:
-                    // cvxeth -> tricrypto2 -> tusd -> susd (cvxeth -> tricrypto2 -> tusd instead)
```
Comment (DanielSchiavini) [inline src/router.ts:434]: @Macket if we are searching for `tusd` we would never search further anyway? :thinking: 
I'm not sure what's the idea behind this

File: src/router.ts [hunk 3]
```diff
@@ -1,459 +1,85 @@
...
-        if (inCoin === outputCoinAddress) {
-            targetRoutes.push(route);
-        } else if (route.route.length < 5) {
-            for (const outCoin in routerGraph[inCoin]) {
-                if (_isVisitedCoin(outCoin, route)) continue;
-
-                for (const step of routerGraph[inCoin][outCoin]) {
-                    const poolData = ALL_POOLS[step.poolId];
-
-                    if (!poolData?.is_lending && _isVisitedPool(step.poolId, route)) continue;
```
Comment (DanielSchiavini) [inline src/router.ts:425]: @Macket does this mean that a lending pool may be visited more than once in the route?

**Review by Macket — APPROVED** [review]

File: src/router.ts (hunk 3 above)
Comment (Macket) [reply src/router.ts:425]: Yes! It's needed for such swap

image

File: src/router.ts (hunk 2 above)
Comment (Macket) [reply src/router.ts:434]: We allow to visit a lending pool more than once in the route, but don't allow this for other pools.

**Pull request merged**